cgen: complex behavior when use reference + generics + or block + val.$(field.name) != unsafe { nil }

### Describe the bug

Maybe a cast problem? 
```v
		if val.$(field.name) != unsafe { nil } {
			$if field.indirections > 0 {
				println('aaaaaaaaaaaaaaaaaaaa')
			} $else {
				println('bbbbbbbbbbbbbbbbbbbb')
			}
		}
```
beacame
```c
		if (!string__eq(*val.val, ((void*)0))) {
			#if 1
			{
				println(_SLIT("aaaaaaaaaaaaaaaaaaaa"));
			}
			#endif
		}
```

### Expected Behavior

print `aaaaaaaaaaaaaaaaaaaa`

### Current Behavior

```bash
==================
/tmp/v_1000/a.12604891049747056293.tmp.c:12542: error: cannot convert 'void *' to 'struct string'
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang
```

### Reproduction Steps

```v
struct Encoder {}
struct Writer {}

struct StructTypePointer[T] {
mut:
	val &T
}

pub fn (e &Encoder) encode_value[T](val T, mut wr Writer) ! {
	e.encode_struct[T](val, 1, mut wr)!
}

fn (e &Encoder) encode_struct[U](val U,level int, mut wr Writer) ! {
	$for field in U.fields {
		if val.$(field.name) != unsafe { nil } {
			$if field.indirections > 0 {
				println('aaaaaaaaaaaaaaaaaaaa')
			} $else {
				println('bbbbbbbbbbbbbbbbbbbb')
			}
		}
	}
}

fn main() {
	e := Encoder{}
	mut sb := Writer{}

	mut string_initialized_with_reference := "ads"

	e.encode_value(StructTypePointer[string]{ val: &string_initialized_with_reference }, mut sb) or { }
}
```

### Possible Solution

_No response_

### Additional Information/Context

_No response_

### V version

V 0.3.3 6f85384

### Environment details (OS name and version, etc.)

Ubuntu

--- a/vlib/v/gen/c/infix.v
+++ b/vlib/v/gen/c/infix.v
@@ -100,7 +100,9 @@ fn (mut g Gen) infix_expr_eq_op(node ast.InfixExpr) {
 	is_none_check := node.left_type.has_flag(.option) && node.right is ast.None
 	if is_none_check {
 		g.gen_is_none_check(node)
-	} else if (left.typ.is_ptr() && right.typ.is_int()) || (right.typ.is_ptr() && left.typ.is_int()) {
+	} else if (left.typ.is_ptr() && right.typ.is_int())
+		|| (right.typ.is_ptr() && left.typ.is_int())
+		|| (left.typ.is_ptr() && right.typ == ast.nil_type) {
 		g.gen_plain_infix_expr(node)
 	} else if (left.typ.idx() == ast.string_type_idx || (!has_defined_eq_operator
 		&& left.unaliased.idx() == ast.string_type_idx)) && node.right is ast.StringLiteral

--- a/vlib/v/tests/comptime_indirection_check_test.v
+++ b/vlib/v/tests/comptime_indirection_check_test.v
@@ -1,0 +1,35 @@
+struct Encoder {}
+
+struct Writer {}
+
+struct StructTypePointer[T] {
+mut:
+	val &T
+}
+
+pub fn (e &Encoder) encode_value[T](val T, mut wr Writer) ! {
+	assert e.encode_struct[T](val, 1, mut wr)! == 'a'
+}
+
+fn (e &Encoder) encode_struct[U](val U, level int, mut wr Writer) !string {
+	$for field in U.fields {
+		if val.$(field.name) != unsafe { nil } {
+			$if field.indirections > 0 {
+				assert field.indirections == 1
+				return 'a'
+			} $else {
+				return 'b'
+			}
+		}
+	}
+	return 'z'
+}
+
+fn test_check() {
+	e := Encoder{}
+	mut sb := Writer{}
+
+	mut string_initialized_with_reference := 'ads'
+	e.encode_value(StructTypePointer[string]{ val: &string_initialized_with_reference }, mut
+		sb) or {}
+}

